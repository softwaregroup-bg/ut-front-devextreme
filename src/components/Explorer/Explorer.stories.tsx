import React from 'react';
import { withReadme } from 'storybook-readme';
import type { Story } from '@storybook/react';
import {Toast} from '../prime';

// @ts-ignore: md file and not a module
import README from './README.md';
import Explorer from './index';
import {fetchItems, updateItems} from './mock';

export default {
    title: 'Explorer',
    component: Explorer,
    decorators: [withReadme(README)],
    args: {
        state: {}
    }
};

const Template: Story<{
    createPermission?: string,
    editPermission?: string,
    deletePermission?: string
}> = ({
    createPermission,
    editPermission,
    deletePermission
}) => {
    const toast = React.useRef(null);
    const show = action => data => toast.current.show({
        severity: 'success',
        summary: 'Submit',
        detail: <pre>{JSON.stringify({action, data}, null, 2)}</pre>
    });
    return (
        <>
            <Toast ref={toast} />
            <div style={{height: 500, display: 'flex', flexDirection: 'column'}}>
                <Explorer
                    fetch={fetchItems}
                    keyField='id'
                    resultSet='items'
                    schema={{
                        properties: {
                            name: {
                                title: 'Name',
                                filter: true,
                                sort: true
                            },
                            size: {
                                title: 'Size',
                                filter: true,
                                sort: true
                            }
                        }
                    }}
                    columns = {['name', 'size']}
                    subscribe={updateItems}
                    details={{
                        name: 'Name'
                    }}
                    filter={{}}
                    actions={[{
                        title: 'Create',
                        permission: createPermission,
                        action: () => {}
                    }, {
                        title: 'Edit',
                        permission: editPermission,
                        enabled: 'current',
                        action: show('edit')
                    }, {
                        title: 'Delete',
                        permission: deletePermission,
                        enabled: 'selected',
                        action: show('delete')
                    }]}
                >
                    <div>Navigation component</div>
                </Explorer>
            </div>
        </>
    );
};

export const Basic = Template.bind({});

export const ActionPermissions = Template.bind({});
ActionPermissions.args = {
    createPermission: 'forbidden',
    editPermission: 'granted',
    deletePermission: 'forbidden'
};
