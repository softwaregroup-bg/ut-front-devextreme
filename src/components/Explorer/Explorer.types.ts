import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';

interface column {
    field: string;
    title: string;
    filter?: boolean;
    action?: ({
        id: any,
        current: {},
        selected: []
    }) => void;
}

interface action {
    title: string;
    permission?: string;
    enabled?: string | boolean;
    action: ({
        id: any,
        current: {},
        selected: []
    }) => void;
}

export interface Props {
    keyField: string;
    resultSet?: string;
    fields: column[];
    fetch: (params: {}) => Promise<{}>;
    className?: string;
    details: {};
    actions?: action[];
    filter?: {};
}

const styles = createStyles({
    details: {
        marginBottom: 15
    },
    detailsLabel: {},
    detailsValue: {}
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
