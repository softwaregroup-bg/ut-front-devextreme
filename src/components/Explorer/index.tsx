import React from 'react';
import lodashGet from 'lodash.get';
import merge from 'ut-function.merge';
import clsx from 'clsx';

import { DataTable, Column, Button, Toolbar, Splitter, SplitterPanel } from '../prime';
import Permission from '../Permission';
import useToggle from '../hooks/useToggle';
import columnProps, {TableFilter} from '../lib/column';

import { Styled, StyledType } from './Explorer.types';

const flexGrow = {flexGrow: 1};
const selectionWidth = {width: '3em'};
const backgroundNone = {background: 'none'};

const Explorer: StyledType = ({
    classes,
    className,
    keyField,
    fetch,
    subscribe,
    schema: {properties = {}} = {},
    columns,
    resultSet,
    children,
    details,
    actions,
    filter,
    onDropdown,
    showFilter = true,
    pageSize = 10
}) => {
    const [tableFilter, setFilters] = React.useState<TableFilter>({
        filters: columns?.reduce((prev, column) => {
            const value = lodashGet(filter, column);
            column = column.split('.').pop();
            return (value === undefined) ? prev : {...prev, [column]: {value}};
        }, {}),
        first: 0,
        page: 1
    });
    const handleFilterPageSort = React.useCallback(event => setFilters(prev => ({...prev, ...event})), []);

    const [selected, setSelected] = React.useState(null);
    const handleSelectionChange = React.useCallback(e => setSelected(e.value), []);

    const [current, setCurrent] = React.useState(null);
    const handleRowSelect = React.useCallback(e => setCurrent(e.data), []);

    const [navigationOpened, navigationToggle] = useToggle(true);
    const [detailsOpened, detailsToggle] = useToggle(true);

    const [loading, setLoading] = React.useState(false);
    const [[items, totalRecords], setItems] = React.useState([[], 0]);
    const [dropdowns, setDropdown] = React.useState({});

    const dropdownNames = (columns || [])
        .flat()
        .filter(Boolean)
        .map(name => lodashGet(properties, name?.replace(/\./g, '.properties.'))?.widget?.dropdown)
        .filter(Boolean)
        .join(',');

    const buttons = React.useMemo(() => (actions || []).map(({title, action, enabled = true, permission}, index) => {
        const isEnabled = enabled => {
            if (typeof enabled !== 'string') return !!enabled;
            switch (enabled) {
                case 'current': return !!current;
                case 'selected': return selected && selected.length > 0;
                default: return false;
            }
        };
        return (
            <Permission permission={permission}>
                <Button
                    key={index}
                    label={title}
                    onClick={() => action({
                        id: current && current[keyField],
                        current,
                        selected
                    })}
                    disabled={!isEnabled(enabled)}
                    className="mr-2"
                />
            </Permission>
        );
    }
    ), [keyField, actions, current, selected]);
    React.useEffect(() => {
        async function load() {
            if (!fetch) {
                setItems([[], 0]);
                setDropdown({});
            } else {
                setLoading(true);
                try {
                    const items = await fetch(merge(
                        {...filter},
                        {
                            [resultSet || 'filterBy']: Object.entries(tableFilter.filters).reduce((prev, [name, {value}]) => ({...prev, [name]: value}), {})
                        },
                        tableFilter.sortField && {
                            orderBy: [{
                                field: tableFilter.sortField,
                                dir: {[-1]: 'DESC', 1: 'ASC'}[tableFilter.sortOrder]
                            }]
                        },
                        pageSize && {
                            paging: {
                                pageSize,
                                pageNumber: Math.floor(tableFilter.first / pageSize) + 1
                            }
                        }
                    ));
                    const records = resultSet ? items[resultSet] : items;
                    let total = items.pagination && items.pagination.recordsTotal;
                    if (total == null) {
                        total = (records && records.length) || 0;
                        if (total === pageSize) total++;
                        total = tableFilter.first + total;
                    }
                    setItems([records, total]);
                    if (onDropdown) setDropdown(await onDropdown(dropdownNames.split(',').filter(Boolean)));
                } finally {
                    setLoading(false);
                }
            }
        }
        load();
        if (subscribe) {
            return subscribe(rows => {
                setItems(([items, totalRecords]) => [(Array.isArray(rows) || !keyField) ? rows : items.map(item => {
                    const update = rows[item[keyField]];
                    return update ? {...item, ...update} : item;
                }), totalRecords]);
            });
        }
    }, [fetch, tableFilter, filter, subscribe, resultSet, pageSize, onDropdown, keyField, dropdownNames]);
    const detailsPanel = React.useMemo(() => detailsOpened && details &&
        <SplitterPanel key='details' size={10}>
            <div style={{ width: '200px' }}>{
                current && Object.entries(details).map(([name, value], index) =>
                    <div className={classes.details} key={index}>
                        <div className={classes.detailsLabel}>{value}</div>
                        <div className={classes.detailsValue}>{current[name]}</div>
                    </div>
                )
            }</div>
        </SplitterPanel>, [classes.details, classes.detailsLabel, classes.detailsValue, current, details, detailsOpened]);

    const filterBy = (name: string, value: string) => e => {
        setFilters(prev => {
            const next = {
                ...prev,
                filters: {
                    ...prev?.filters,
                    [name]: {
                        ...prev?.filters?.[name],
                        value: e[value]
                    }
                }
            };
            return next;
        });
    };

    const Columns = React.useMemo(() => columns.map(name => {
        const property = lodashGet(properties, name?.replace(/\./g, '.properties.'));
        const field = name.split('.').pop();
        return (
            <Column
                key={name}
                field={field}
                header={property?.title}
                body={property?.action && (row => <Button
                    label={row[field]}
                    style={{padding: 0, minWidth: 'inherit'}}
                    className='p-button-link'
                    onClick={() => property.action({
                        id: row && row[keyField],
                        current: row,
                        selected: [row]
                    })}
                />)}
                filter={showFilter && !!property?.filter}
                sortable={!!property?.sort}
                {...columnProps({name: field, property, dropdowns, tableFilter, filterBy})}
            />
        );
    }), [columns, properties, showFilter, dropdowns, tableFilter, keyField]);
    const hasChildren = !!children;
    const left = React.useMemo(() => <>
        {hasChildren && <Button icon="pi pi-bars" className="mr-2" onClick={navigationToggle}/>}
        {buttons}
    </>, [navigationToggle, buttons, hasChildren]);
    const right = React.useMemo(() => <Button icon="pi pi-bars" className="mr-2" onClick={detailsToggle}/>, [detailsToggle]);

    const table = <DataTable
        autoLayout
        lazy
        rows={pageSize}
        totalRecords={totalRecords}
        paginator
        first={tableFilter.first}
        sortField={tableFilter.sortField}
        sortOrder={tableFilter.sortOrder}
        filters={tableFilter.filters}
        onPage={handleFilterPageSort}
        onSort={handleFilterPageSort}
        onFilter={handleFilterPageSort}
        loading={loading}
        dataKey={keyField}
        value={items}
        selection={selected}
        onSelectionChange={handleSelectionChange}
        onRowSelect={handleRowSelect}
    >
        {keyField && <Column selectionMode="multiple" style={selectionWidth}/>}
        {Columns}
    </DataTable>;
    return (
        <div className={clsx('flex', 'flex-column', className)} style={{height: '100%'}} >
            {(detailsPanel || children || buttons?.length) ? <>
                <Toolbar left={left} right={right} style={backgroundNone} />
                <Splitter style={flexGrow}>
                    {[
                        children && navigationOpened && <SplitterPanel key='nav' size={15}>{children}</SplitterPanel>,
                        <SplitterPanel key='items' size={75}>
                            {table}
                        </SplitterPanel>,
                        detailsPanel
                    ].filter(Boolean)}
                </Splitter>
            </> : table}
        </div>
    );
};

export default Styled(Explorer);
