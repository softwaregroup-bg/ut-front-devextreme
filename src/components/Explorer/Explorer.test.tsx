import React from 'react';
import {waitFor} from '@testing-library/react';

import { render } from '../test';
import Explorer from './index';

describe('<Explorer />', () => {
    it('render equals snapshot', async() => {
        const { getByTestId } = render(<Explorer
            fetch={() => Promise.resolve({
                items: [...Array(50).keys()].map(number => ({
                    id: number,
                    name: `Item ${number}`,
                    size: number * 10
                }))
            })}
            keyField='id'
            resultSet='items'
            fields={[{
                field: 'name',
                title: 'Name',
                filter: true
            }, {
                field: 'size',
                title: 'Size'
            }]}
            details={{
                name: 'Name'
            }}
        />);
        await waitFor(() => expect(getByTestId('ut-front-test').querySelector('tr.p-selectable-row')).toBeTruthy());
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
