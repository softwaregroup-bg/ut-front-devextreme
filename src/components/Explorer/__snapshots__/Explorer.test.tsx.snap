// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Explorer /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="p-d-flex p-flex-column"
    style="height: 100%;"
  >
    <div
      class="p-toolbar p-component"
      role="toolbar"
    >
      <div
        class="p-toolbar-group-left"
      >
        <button
          class="p-button p-component p-mr-2 p-button-icon-only"
        >
          <span
            class="p-button-icon p-c pi pi-bars"
          />
          <span
            class="p-button-label p-c"
          >
             
          </span>
        </button>
      </div>
      <div
        class="p-toolbar-group-right"
      >
        <button
          class="p-button p-component p-mr-2 p-button-icon-only"
        >
          <span
            class="p-button-icon p-c pi pi-bars"
          />
          <span
            class="p-button-label p-c"
          >
             
          </span>
        </button>
      </div>
    </div>
    <div
      class="p-splitter p-component p-splitter-horizontal"
      style="flex-grow: 1;"
    >
      <div
        class="p-splitter-panel"
        style="flex-basis: calc(15% - 8px);"
      />
      <div
        class="p-splitter-gutter"
        style="width: 4px;"
      >
        <div
          class="p-splitter-gutter-handle"
        />
      </div>
      <div
        class="p-splitter-panel"
        style="flex-basis: calc(75% - 8px);"
      >
        <div
          class="p-datatable p-component p-datatable-auto-layout p-datatable-hoverable-rows"
          data-scrollselectors=".p-datatable-scrollable-body, .p-datatable-unfrozen-view .p-datatable-scrollable-body"
        >
          <div
            class="p-datatable-wrapper"
          >
            <table
              role="grid"
            >
              <thead
                class="p-datatable-thead"
              >
                <tr
                  role="row"
                >
                  <th
                    class="p-sortable-disabled p-selection-column"
                    role="columnheader"
                    style="width: 3em;"
                  >
                    <span
                      class="p-column-title"
                    />
                  </th>
                  <th
                    class="p-sortable-disabled"
                    role="columnheader"
                  >
                    <span
                      class="p-column-title"
                    >
                      Name
                    </span>
                  </th>
                  <th
                    class="p-sortable-disabled"
                    role="columnheader"
                  >
                    <span
                      class="p-column-title"
                    >
                      Size
                    </span>
                  </th>
                </tr>
                <tr
                  role="row"
                >
                  <th
                    class="p-filter-column"
                    role="columnheader"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </th>
                  <th
                    class="p-filter-column"
                    role="columnheader"
                  >
                    <input
                      class="p-inputtext p-component"
                      placeholder="Filter by Name"
                      type="text"
                      value=""
                    />
                  </th>
                  <th
                    class="p-filter-column"
                    role="columnheader"
                  />
                </tr>
              </thead>
              <tbody
                class="p-datatable-tbody"
              >
                <tr
                  class="p-selectable-row"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 0
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    0
                  </td>
                </tr>
                <tr
                  class="p-selectable-row p-row-odd"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 1
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    10
                  </td>
                </tr>
                <tr
                  class="p-selectable-row"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 2
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    20
                  </td>
                </tr>
                <tr
                  class="p-selectable-row p-row-odd"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 3
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    30
                  </td>
                </tr>
                <tr
                  class="p-selectable-row"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 4
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    40
                  </td>
                </tr>
                <tr
                  class="p-selectable-row p-row-odd"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 5
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    50
                  </td>
                </tr>
                <tr
                  class="p-selectable-row"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 6
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    60
                  </td>
                </tr>
                <tr
                  class="p-selectable-row p-row-odd"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 7
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    70
                  </td>
                </tr>
                <tr
                  class="p-selectable-row"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 8
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    80
                  </td>
                </tr>
                <tr
                  class="p-selectable-row p-row-odd"
                  role="row"
                >
                  <td
                    class="p-selection-column"
                    role="cell"
                    style="width: 3em;"
                  >
                    <div
                      class="p-checkbox p-component"
                    >
                      <div
                        aria-checked="false"
                        class="p-checkbox-box p-component p-clickable"
                        role="checkbox"
                        tabindex="0"
                      >
                        <span
                          class="p-checkbox-icon p-clickable"
                        />
                      </div>
                    </div>
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    Item 9
                  </td>
                  <td
                    class=""
                    role="cell"
                  >
                    90
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="p-paginator p-component p-paginator-bottom"
          >
            <button
              class="p-paginator-first p-paginator-element p-link p-disabled"
              disabled=""
              type="button"
            >
              <span
                class="p-paginator-icon pi pi-angle-double-left"
              />
            </button>
            <button
              class="p-paginator-prev p-paginator-element p-link p-disabled"
              disabled=""
              type="button"
            >
              <span
                class="p-paginator-icon pi pi-angle-left"
              />
            </button>
            <span
              class="p-paginator-pages"
            >
              <button
                class="p-paginator-page p-paginator-element p-link p-paginator-page-start p-highlight"
                type="button"
              >
                1
              </button>
              <button
                class="p-paginator-page p-paginator-element p-link"
                type="button"
              >
                2
              </button>
              <button
                class="p-paginator-page p-paginator-element p-link"
                type="button"
              >
                3
              </button>
              <button
                class="p-paginator-page p-paginator-element p-link"
                type="button"
              >
                4
              </button>
              <button
                class="p-paginator-page p-paginator-element p-link p-paginator-page-end"
                type="button"
              >
                5
              </button>
            </span>
            <button
              class="p-paginator-next p-paginator-element p-link"
              type="button"
            >
              <span
                class="p-paginator-icon pi pi-angle-right"
              />
            </button>
            <button
              class="p-paginator-last p-paginator-element p-link"
              type="button"
            >
              <span
                class="p-paginator-icon pi pi-angle-double-right"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="p-splitter-gutter"
        style="width: 4px;"
      >
        <div
          class="p-splitter-gutter-handle"
        />
      </div>
      <div
        class="p-splitter-panel p-splitter-panel-nested"
        style="flex-basis: calc(10% - 8px);"
      >
        <div
          style="width: 200px;"
        />
      </div>
    </div>
  </div>
</div>
`;
